test/computational_graph.jl

```diff
@@ -117,83 +115,106 @@ end
-        @test gg5.subgraph_factors == [1, 1, 1]
-        standardize_order!(gg5)
-        @test external(gg5) == OperatorProduct(V5)[[6, 1, 8]]
-        @test gg5.factor ≈ g5.factor
-        # gg5 = feynman_diagram(V5, [1, 2, 1, 3, 3, 2], external=[1, 2, 3])
-        ggg5 = feynman_diagram(V5, [[2, 4], [3, 9], [5, 7]])
-        @test isequiv(g5, ggg5, :id, :type)
+        V5p = [𝑓⁺(11, true), 𝑓⁻(1, true), 𝜙(12, true), 𝑓⁺(2)𝑓⁻(3)𝜙(4), 𝑓⁺(5)𝑓⁻(6)𝜙(7), 𝑓⁺(8)𝑓⁻(9)𝜙(10)]
+        g5p = feynman_diagram(V5p, [[5, 7], [6, 12], [8, 10], [2, 4], [3, 9], [11, 1]], external=[11, 9, 4])
+        @test g5.factor ≈ g5p.factor    # reorder of external fake legs will not change the sign.
+        @test g5p.subgraph_factors == [1, 1, 1]
+        standardize_order!(g5p)
+        @test external(g5p) == [𝑓⁻(9), 𝜙(7), 𝑓⁺(2)]
+        @test g5p.factor ≈ g5.factor
 
-        V6 = [𝑓⁺(1), 𝑓⁺(2)𝑓⁻(3)𝜙(4), 𝑓⁺(5)𝑓⁻(6)𝜙(7), 𝑓⁻(8)]
-        g6 = feynman_diagram(V6, [[1, 3], [2, 6], [4, 7], [5, 8]], external=[8, 1])    # fermionic Green2
+        V6 = [𝑓⁻(8), 𝑓⁺(1), 𝑓⁺(2)𝑓⁻(3)𝜙(4), 𝑓⁺(5)𝑓⁻(6)𝜙(7)]
+        g6 = feynman_diagram(V6, [[2, 4], [3, 7], [5, 8], [6, 1]], external=[1, 2])    # fermionic Green2
         @test g6.factor == -1
         @test g6.subgraph_factors == [1, 1, 1, 1]
-        @test external(g6) == OperatorProduct(V6)[[8, 1]]
+        @test external(g6) == [𝑓⁻(8), 𝑓⁺(1)]
         standardize_order!(g6)
         @test g6.factor == 1
-        @test external(g6) == OperatorProduct(V6)[[1, 8]]
+        @test vertices(g6)[1:2] == [𝑓⁺(1), 𝑓⁻(8)]
+        @test Set(external(g6)) == Set([𝑓⁺(1), 𝑓⁻(8)])
 
-        V7 = [𝑓⁺(1)𝑓⁻(2)𝜙(3), 𝑓⁺(4)𝑓⁻(5)𝜙(6), 𝑓⁻(7)]
-        g7 = feynman_diagram(V7, [[1, 5], [3, 6], [4, 7]], external=[2, 7])     # sigma*G
+        V7 = [𝑓⁻(7), 𝑓⁺(8, true), 𝑓⁺(1)𝑓⁻(2)𝜙(3), 𝑓⁺(4)𝑓⁻(5)𝜙(6)]
+        g7 = feynman_diagram(V7, [[3, 7], [5, 8], [6, 1], [4, 2]], external=[1, 4])     # sigma*G
         @test g7.factor == 1
+        @test external(g7) == [𝑓⁻(7), 𝑓⁻(2)]
+        @test external_with_ghost(g7) == [𝑓⁻(7), 𝑓⁺(8, true)]
 
-        V8 = [𝑓⁺(1), 𝑓⁺(2)𝑓⁻(3)𝜙(4), 𝑓⁺(5)𝑓⁻(6)𝜙(7), 𝑓⁺(8)𝑓⁻(9)𝜙(10), 𝑓⁻(11), 𝑓⁺(12)𝑓⁻(13)𝜙(14)]
-        g8 = feynman_diagram(V8, [[1, 3], [2, 6], [4, 14], [5, 13], [7, 10], [8, 11]], external=[1, 9, 11, 12])
+        V8 = [𝑓⁻(1, true), 𝑓⁺(2), 𝑓⁻(12), 𝑓⁺(16, true), 𝑓⁺(3)𝑓⁻(4)𝜙(5), 𝑓⁺(6)𝑓⁻(7)𝜙(8), 𝑓⁺(9)𝑓⁻(10)𝜙(11), 𝑓⁺(13)𝑓⁻(14)𝜙(15)]
+        g8 = feynman_diagram(V8, [[2, 6], [5, 9], [7, 16], [8, 15], [10, 13], [11, 3], [12, 4], [14, 1]], external=[2, 12, 3, 14])
         @test g8.factor == -1
-        gg8 = feynman_diagram(V8, [[1, 3], [2, 6], [4, 14], [5, 13], [7, 10], [8, 11]], external=[9, 1, 11, 12])
-        @test gg8.factor == -1
+        @test external(g8) == [𝑓⁺(2), 𝑓⁻(10), 𝑓⁻(12), 𝑓⁺(13)]
+        @test external_with_ghost(g8) == [𝑓⁻(1, true), 𝑓⁺(2), 𝑓⁻(12), 𝑓⁺(16, true)]
+        standardize_order!(g8)
+        @test external_with_ghost(g8) == [𝑓⁺(2), 𝑓⁺(16, true), 𝑓⁻(12), 𝑓⁻(1, true)]
```

Comment:
Why is external(g8) ⊈ external_with_ghost(g8)? I wonder if it is a bug in `external_with_ghost`.

Reply:
Kun explained the conceptual misunderstanding to me; it is not a bug.